sw/device/lib/dif/dif_hmac.h

```diff
@@ -0,0 +1,211 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 1.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef SW_DEVICE_LIB_DIF_HMAC_DEVICE_H_
+#define SW_DEVICE_LIB_DIF_HMAC_DEVICE_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+
+/**
+ * Supported HMAC modes of operation.
+ */
+typedef enum dif_hmac_mode {
+  /** The HMAC mode */
+  kDifHmacModeHmac = 0,
+  /** The SHA256-only mode */
+  kDifHmacModeSha256,
+} dif_hmac_mode_t;
+
+/**
+ * Supported byte endienness options.
+ */
+typedef enum dif_hmac_endianness {
+  /** Big endian byte ordering */
+  kDifHmacBigEndian = 0,
+  /** Little endian byte ordering */
+  kDifHmacLittleEndian,
+} dif_hmac_endianness_t;
+
+/**
+ * Error codes for HMAC functions that may exhibit generic failures.
+ */
+typedef enum dif_hmac_error {
+  /** No error occurred. */
+  kDifHmacNoError = 0,
+  /** An invalid argument was provided. */
+  kDifHmacInvalidArgument,
+} dif_hmac_error_t;
+
+/**
+ * Error codes for HMAC FIFO operations that may fail.
+ */
+typedef enum dif_hmac_fifo_error {
+  /** No error occurred. */
+  kDifHmacFifoNoError = 0,
+  /** An invalid argument was provided. */
+  kDifHmacFifoInvalidArgument,
+  /** The FIFO filled up before the buffer was fully consumed. */
+  kDifHmacFifoFull,
+} dif_hmac_fifo_error_t;
+
+/**
+ * Configuration for initializing the HMAC device.
+ */
+typedef struct dif_hmac_config {
+  /** The base address for registers in the HMAC IP*/
+  mmio_region_t base_addr;
+  /** The mode of operation */
+  dif_hmac_mode_t mode;
+  /** Byte endianness for writes to the FIFO */
+  dif_hmac_endianness_t message_endianness;
+  /** Byte endianness for reads from the digest */
+  dif_hmac_endianness_t digest_endianness;
+  /** The HMAC key. Only used in |kDifHmacModeHmac| mode. */
+  uint32_t key[8];
```

Comment:
Is there a reason this can't be a pointer instead? It feels somehow gross to me to pass a key inline.

Also, I'm somewhat worried about this being an array of words, rather than an array of bytes. This puts an alignment requirement on keys callers are passing in, which may result in unnecessary copies.

Reply:
Since the key is optional in SHA256 mode I was trying to avoid a garbage pointer making its way in here. I can make it a pointer if that's preferable.